metronome/hotstuff/consensus/src/metronome/hotstuff/consensus/basic/ProtocolState.scala

```diff
@@ -0,0 +1,492 @@
+
+  /** Check that the proposed new block extends the locked Q.C. (safety)
+    * or that the Quorum Certificate is newer than the locked Q.C. (liveness).
+    */
+  private def isSafe(m: Prepare[A]): Boolean = {
+    val valid = isExtension(m.block, m.highQC)
+    val safe  = isExtension(m.block, lockedQC)
+    val live  = m.highQC.viewNumber > lockedQC.viewNumber
+
+    valid && (safe || live)
+  }
+
+  /** Check that a block extends from the one in the Q.C.
+    *
+    * Currently only allows direct parent-child relationship,
+    * which means each leader is expected to create max 1 block
+    * on top of the previous high Q.C.
+    */
+  private def isExtension(block: A#Block, qc: QuorumCertificate[A]): Boolean =
+    qc.blockHash == Block[A].parentBlockHash(block)
+
+  /** Register a new vote; if there are enough to form a new Q.C.,
+    * do so and broadcast it.
+    */
+  private def addVoteAndMaybeBroadcastQC(vote: Vote[A]): Transition[A] = {
+    // `matchingVote` made sure all votes are for the same content,
+    // and `moveTo` clears the votes, so they should be uniform.
+    val next = copy(votes = votes + vote)
+
+    // Only make the quorum certificate once.
+    val effects =
+      if (votes.size < quorumSize && next.votes.size == quorumSize) {
+        val vs = next.votes.toSeq
+        val qc = QuorumCertificate(
+          phase = vs.head.phase,
+          viewNumber = vs.head.viewNumber,
+          blockHash = vs.head.blockHash,
+          signature = Signing[A].combine(vs.map(_.signature))
+        )
+        broadcast {
+          Quorum(viewNumber, qc)
+        }
+      } else Nil
+
+    // The move to the next phase will be triggered when the Q.C. is delivered.
+    next -> effects
+  }
+
+  /** Register a NewView from a replica; if there are enough, select the High Q.C. and create a block. */
+  private def addNewViewAndMaybeCreateBlock(
+      sender: A#PKey,
+      newView: NewView[A]
+  ): Transition[A] = {
+    // We already checked that these are for the current view.
+    val next = copy(newViews =
+      newViews.updated(
+        sender,
+        newViews.get(sender).fold(newView) { oldView =>
+          if (newView.prepareQC.viewNumber > oldView.prepareQC.viewNumber)
+            newView
+          else oldView
+        }
+      )
+    )
+
+    // Only make a block once.
+    val effects =
+      if (newViews.size < quorumSize && next.newViews.size == quorumSize) {
+        List(
+          CreateBlock(
+            viewNumber,
+            highQC = next.newViews.values.map(_.prepareQC).maxBy(_.viewNumber)
+          )
+        )
+      } else Nil
+
+    // The move to the next phase will be triggered when the block is created.
+    next -> effects
+  }
+}
+
+object ProtocolState {
+
+  /** The result of state transitions are the next state and some effects
+    * that can be carried out in parallel.
+    */
+  type Transition[A <: Agreement] = (ProtocolState[A], Seq[Effect[A]])
```

Comment:
I don't have much experience coding with Scala and I am curious about the architecture for Hotstuff in general. Based on the code here, in my opinion, there are two threads running in parallel with one running the protocolState update and the other handling effects in sequence. If that's the case, how to process some effects like ScheduleNextView, which correlates with protocolState update as well, and shall we do it synchronously or asynchronously.

Reply:
Thank you for the question! I tried to describe on a high level under the "Host System" section of the PR description.

The idea is that there would be 1 fiber going through the `Event`s in a `ConcurrentQueue`, where every event can create a new `ProtocolState` and some `Effect`s. The effects can be handled concurrently, they don't have to be in sequence; for example we can send messages to all federation members in parallel. If an `Effect` results in a new `Event` (such as a `CreateBlock` eventually produces a `BlockCreated`), that goes in the event queue, to be processed in sequence. Similarly, the host system can carry out `ScheduleNextView` in the background and put a `NextView` event on the queue when its' done:

```scala
def startProcessingEffect(effect: Effect): F[Fiber[F, Unit]] = effect match {
  case ScheduleNextView(viewNumber, timeout) = 
    Concurrent[F].start {
      Timer[F].sleep(timeout) >> eventQueue.offer(Event(NextView(viewNumber))
    }
  ...
}
```

Here's a [blog](https://dev.to/kspeakman/testable-back-end-programming-in-f-41n5) that illustrates this design beautifully.

Reply:
Nice introduction. I will take a look at the blog.

Reply:
Here are links to the classes mentioned: 
* [ConcurrentQueue](https://monix.io/api/3.1/monix/catnap/ConcurrentQueue.html)
* [Ref](https://typelevel.org/cats-effect/concurrency/ref.html)
* [Timer](https://typelevel.org/cats-effect/datatypes/timer.html)
* [Fiber](https://typelevel.org/cats-effect/datatypes/fiber.html)
* [Concurrent](https://typelevel.org/cats-effect/typeclasses/concurrent.html)
* [cats-effect](https://typelevel.org/cats-effect/api/cats/effect/index.html) API docs in general